### PR TITLE
combining window and summary for memory

### PR DESCRIPTION
**Pull request by j-csc**
Open to feedbacks. Will add in support for summarizing once reaching token limit

### CODE REVIEW

File: langchain/chains/conversation/memory.py [hunk 1]
```diff
@@ -130,3 +130,70 @@ def save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None:
...
+        ai = f"Assistant: {list(outputs.values())[0]}"
+        new_lines = "\n".join([human, ai])
+        self.buffer.append(new_lines)
+        if len(self.buffer) > self.k:
+            pruned_memory = self.buffer[:-self.k]
+            chain = LLMChain(llm=self.llm, prompt=self.prompt)
+            self.moving_summary_buffer = chain.predict(summary=self.moving_summary_buffer, new_lines=("\n".join(pruned_memory)))
+            self.buffer = self.buffer[-5:]
+
+    def restore_memory_variables(self, _moving_summary_buffer: str, _buffer: List[str]) -> None:
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:191]: why is this neccessary as opposed to just initializing the memory class with these variables? not against it just trying to understand.

File: langchain/chains/conversation/memory.py [hunk 2]
```diff
@@ -130,3 +130,70 @@ def save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None:
...
+            raise ValueError(f"One output key expected, got {outputs.keys()}")
+        human = f"Human: {inputs[prompt_input_key]}"
+        ai = f"Assistant: {list(outputs.values())[0]}"
+        new_lines = "\n".join([human, ai])
+        self.buffer.append(new_lines)
+        if len(self.buffer) > self.k:
+            pruned_memory = self.buffer[:-self.k]
+            chain = LLMChain(llm=self.llm, prompt=self.prompt)
+            self.moving_summary_buffer = chain.predict(summary=self.moving_summary_buffer, new_lines=("\n".join(pruned_memory)))
+            self.buffer = self.buffer[-5:]
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:189]: should use self.k here, right?

File: langchain/chains/conversation/memory.py (hunk 1 above)
Comment (jaloo555) [reply langchain/chains/conversation/memory.py:191]: ahh will make the change. mainly was using this with my own server where the instance frequently gets lost

File: langchain/chains/conversation/memory.py (hunk 2 above)
Comment (jaloo555) [reply langchain/chains/conversation/memory.py:189]: yep!

**Comment by jaloo555** [comment]
just updated! @hwchase17

**Pull request merged**